File: rd-text/src/main/kotlin/com/jetbrains/rider/rdtext/impl/intrinsics/RdTextBufferChange.kt
package com.jetbrains.rider.rdtext.impl.intrinsics

import com.jetbrains.rider.framework.IMarshaller
import com.jetbrains.rider.framework.SerializationCtx
import com.jetbrains.rider.framework.AbstractBuffer
import com.jetbrains.rider.rdtext.TextBufferVersion
import com.jetbrains.rider.rdtext.intrinsics.RdTextChange
import com.jetbrains.rider.framework.readEnum
import com.jetbrains.rider.framework.writeEnum
import com.jetbrains.rider.rdtext.intrinsics.RdTextChangeMarshaller
import com.jetbrains.rider.util.string.IPrintable
import com.jetbrains.rider.util.string.PrettyPrinter
import com.jetbrains.rider.util.string.printToString
import kotlin.reflect.*

data class RdTextBufferChange(val version: TextBufferVersion, val origin: RdChangeOrigin, val change: RdTextChange): IPrintable {
    companion object : IMarshaller<RdTextBufferChange> {

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): RdTextBufferChange {
            val masterVersionRemote = buffer.readInt()
            val slaveVersionRemote = buffer.readInt()
            val version = TextBufferVersion(masterVersionRemote, slaveVersionRemote)
            val origin = buffer.readEnum<RdChangeOrigin>()
            val change = RdTextChangeMarshaller.read(ctx, buffer)
            return RdTextBufferChange(version, origin, change)
        }

        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: RdTextBufferChange) {
            val version = value.version
            buffer.writeInt(version.master)
            buffer.writeInt(version.slave)
            val origin = value.origin
            buffer.writeEnum(origin)
            val change = value.change
            RdTextChangeMarshaller.write(ctx, buffer, change)
        }

        override val _type: KClass<RdTextBufferChange> get() = RdTextBufferChange::class
    }

    override fun print(printer: PrettyPrinter) {
        printer.println("RdTextBufferChange (")
        printer.indent {
            println("version = (master=${version.master}, slave=${version.slave})")
            println("side = $origin")
            print("change = ")
            change.print(printer)
        }
        printer.println(")")
    }

    override fun toString(): String = this.printToString()
}

File: rd-net/RdFramework/Text/Impl/Intrinsics/RdTextBufferChange.cs
using JetBrains.Rd.Base;
using JetBrains.Rd.Text.Intrinsics;
using JetBrains.Rd.Util;

namespace JetBrains.Rd.Text.Impl.Intrinsics
{
  public class RdTextBufferChange : IPrintable
  {
    public readonly RdTextChange Change;
    public readonly TextBufferVersion Version;
    public readonly RdChangeOrigin Origin;

    public RdTextBufferChange(TextBufferVersion version, RdChangeOrigin origin, RdTextChange change)
    {
      Change = change;
      Version = version;
      Origin = origin;
    }

    public static CtxReadDelegate<RdTextBufferChange> ReadDelegate = (ctx, stream) =>
    {
      var masterVersionRemote = stream.ReadInt();
      var slaveVersionRemote = stream.ReadInt();
      var textBufferVersion = new TextBufferVersion(masterVersionRemote, slaveVersionRemote);
      var side = (RdChangeOrigin)stream.ReadInt();
      var textChange = RdTextChangeSerializer.ReadDelegate(ctx, stream);
      return new RdTextBufferChange(textBufferVersion, side, textChange);
    };

    public static CtxWriteDelegate<RdTextBufferChange> WriteDelegate = (ctx, writer, value) =>
    {
      var version = value.Version;
      writer.WriteInt32(version.Master);
      writer.WriteInt32(version.Slave);
      var side = value.Origin;
      writer.WriteInt32((int)side);
      var change = value.Change;
      RdTextChangeSerializer.WriteDelegate(ctx, writer, change);
    };

    public void Print(PrettyPrinter printer)
    {
      printer.Println("RdTextBufferChange (");
      using (printer.IndentCookie())
      {
        printer.Println($"vesion = (master={Version.Master}, slave={Version.Slave})");
        printer.Print("change = ");
        Change.Print(printer);
      }
      printer.Print(")");
    }
  }
}